### ISSUE TITLE
Unfolding breaks when higherorder flag is on 

### ISSUE THREAD

**Issue by nikivazou** [ucsd-progsys/liquidhaskell#1034]
```
{-@ LIQUID "--higherorder" @-}
module Iso where

data P a b  = P {p1 :: a, p2 :: b}
{-@ data P a b = P {p1 :: a, p2 :: b}@-}

{-@ check :: x:a -> {v:P a b | p1 v == x} -> b @-} 
check :: a -> P a b -> b 
check x (P y p) = p

-- THIS IS SAFE
{-@ ex1 :: P a b -> b @-}
ex1 ::  (P a b) -> b 
ex1 f@(P y _) =
  check y f

-- This is UNSAFE
{-@ ex2 :: P a b -> b @-}
ex2 ::  (P a b) -> b 
ex2 f =
  let (P y p) = f in check y f
```

**Comment by ranjitjhala** [comment]
The issue here is that:

1. `--higherorder` disables quals,

2. we *need* quals in `ex2` because that particular 
    thing generates weird GHC core, corresponding to 
    `ProjectE` that yields a CUT variable,

3. The way to get quals in such situations is to write 
    a type alias, e.g.

```
{-@ type P1X a b X = {v:P a b | p1 v == X} @-}
```

but the scraping mechanism is broken due to the "free" (not actually!) 
`X`, and so `Qualifier.refTopQuals` does not _actually_ generate a 
qualifier for it.

Not sure what the fix is. WOULD be nice to not need any quals here, 
but the `ProjectE` causes some rather nasty blowups IIRC. 

(Perhaps that is not true anymore... will have to double-check.)

**Comment by ranjitjhala** [comment]
Fixed by or #1042
